backend/src/mock/bus_checker.rs

```diff
@@ -148,40 +145,68 @@ impl<'a, F: FieldElement> BusChecker<'a, F> {
 
                             (bus_connection, tuple, multiplicity)
                         })
+                        .collect::<Vec<_>>()
                 })
             })
+            // fold the interactions from a row into a state
             .fold(
-                Default::default(),
-                |mut counts, (bus_connection, tuple, multiplicity)| {
-                    // update the counts
-                    let (s, r) = counts.entry(tuple).or_default();
-
+                BusState::default,
+                |mut state, (bus_connection, tuple, multiplicity)| {
                     let abs = multiplicity.unsigned_abs() as usize;
 
                     match multiplicity.cmp(&0) {
                         // if the multiplicity is zero, we don't need to do anything
                         Ordering::Equal => {}
                         // if the multiplicity is positive, we send
                         Ordering::Greater => {
-                            s.entry(bus_connection)
+                            let TupleState { sends, .. } = state.entry(tuple).or_default();
+                            sends
+                                .entry(bus_connection)
                                 .and_modify(|sends| *sends += abs)
                                 .or_insert(abs);
                         }
                         // if the multiplicity is negative, we receive
                         Ordering::Less => {
-                            r.entry(bus_connection)
+                            let TupleState { receives, .. } = state.entry(tuple).or_default();
+                            receives
+                                .entry(bus_connection)
                                 .and_modify(|receives| *receives += abs)
                                 .or_insert(abs);
                         }
                     }
 
-                    counts
+                    state
                 },
-            );
+            )
+            // combine all the states to one
```

Comment:
```suggestion
            // combine all the states to one,
            // because `ParallelIterator::fold` returns a list of intermediate results.
```

Reply:
not needed now that we're just doing map reduce